### ISSUE TITLE
Analyzer produces incorrect result for "typedef G<X extends A<X>, Y extends X> = X Function(Y)"

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#35068]
Dart SDK Version: 2.1.0-dev.9.1
OS: Windows 10

This issue looks like #34861, however it deals with typedefs instead of classes.
Here is a source code example:
``` dart
typedef F<X> = void Function<Y extends X>();
F<X> toF<X>(X x) => null;

class A<X> {}
typedef G<X extends A<X>, Y extends X> = X Function(Y);

main() {
  G source;
  var fsource = toF(source);
  F<G<A<dynamic>, A<dynamic>>> target2 = fsource;
  F<G<A<dynamic>, dynamic>> target = fsource;     //# 02: compile-time error
}
```
Dart produces compile time error with the line 11 (`F<G<A<dynamic>, dynamic>> target = fsource`) and passes with the line 10 (`F<G<A<dynamic>, A<dynamic>>> target2 = fsource`).

Analyzer shows opposite result: it throws compile time error with line 10 and passes with the line 11.

**Comment by eernstg** [comment]
This is a case where CL 80140 matters: The second type parameter of `G` is contravariant, so i2b on `G` proceeds as follows:
```dart
X                    Y
--------------------------------------------------
A<X>                 X                    // Initial values.
A<dynamic>           X                    // Break cycle {X}.
A<dynamic>           Null                 // Replace contravariant occurrence of X.
```
So `G` is expanded by i2b into `G<A<dynamic>, Null>`, and with the old rules (where we ignore that `X` occurs in a contravariant position in `G<..., X>`) we would replace `X` by the current (and final) value in the `X` column, yielding `G<A<dynamic>, A<dynamic>>`.

So the point is that we used to consider the type arguments in isolation, but with 80140 we consider each type argument with the variance of the corresponding type parameter of `G`. That type parameter is contravariant, so a plain `X` is a contravariant occurrence of `X`, whereas, for instance, `Function(X)` would have a covariant occurrence of `X` (because `Function(X)` is in a contravariant position, so `X` of `Function(X)` is in a covariant position).

**Comment by MichaelRFairhurst** [comment]
@lrhn is this also blocked on spec work?

**Comment by MichaelRFairhurst** [comment]
I think this is a duplicate of #34663. Currently analyzer does not consider co(ntra)variance _anywhere_ in i2b, so any bugs related to that are all related to our need to implement that.

**Comment by eernstg** [comment]
This example differs from #34663 in that the type parameters of `G` have non-trivial variances (they are not all covariant): `X` is covariant and `Y` is contravariant; in #34662, both type parameters of `G` are covariant, and this matters for the relevant algorithms. So this is related to #34663, but it isn't just a duplicate.

**Comment by MichaelRFairhurst** [comment]
True, the examples have differences. However, the real issue is that we implement an old version of I2B. Fixing that will fix both cases....I don't intend on fixing one of these reproes before or after the other, and so from my perspective, they're the same bug.

**Comment by eernstg** [comment]
OK!
